particle-iot/device-os
# CoAP Blockwise transfer for Describe messages [sc-89923]

### Problem

This PR adds initial support for CoAP blockwise transfers (see [RFC 7959](https://datatracker.ietf.org/doc/html/rfc7959)) that allow sending large CoAP requests and responses that would not normally fit into a single UDP packet. For now, blockwise transfer is only implemented for Describe messages.

### Steps to Test

- Flash the example application to a Gen 2 or Gen 3 device and verify that `particle list` shows all the functions and variables registered by the application.
- Apply [this diff](https://github.com/particle-iot/device-os/files/7643767/disable_device_initiated_describe.diff.txt) to disable device-initiated Describe messages, recompile and flash the system firmware and repeat the above test.

### Example App

```c
#include "application.h"

SYSTEM_MODE(SEMI_AUTOMATIC);
SYSTEM_THREAD(ENABLED);

namespace {

int cloudFn(const String& arg) {
    return 0;
}

int cloudVar = 0;

} // namespace

void setup() {
    for (int i = 1; i <= 50; ++i) {
        auto name = String::format("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx_%04d", i);
        Particle.function(name, cloudFn);
    }
    for (int i = 1; i <= 50; ++i) {
        auto name = String::format("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx_%04d", i);
        Particle.variable(name, cloudVar);
    }
    Particle.connect();
}

void loop() {
}
```

### References

- [sc-89923]
